src/main/scala/org/broadinstitute/dsde/workbench/sam/openam/LdapAccessPolicyDAO.scala

```diff
@@ -139,27 +136,34 @@ class LdapAccessPolicyDAO(protected val ldapConnectionPool: LDAPConnectionPool,
     } yield Resource(resourceTypeName, ResourceId(resourceId), authDomains)
   }
 
-  override def listAccessPolicies(resource: Resource): IO[Set[AccessPolicy]] = IO(
+  override def listAccessPolicies(resource: ResourceIdentity): IO[Set[AccessPolicy]] = cs.evalOn(blockingEc)(IO(
     ldapSearchStream(resourceDn(resource), SearchScope.SUB, Filter.createEqualityFilter("objectclass", ObjectClass.policy))(unmarshalAccessPolicy).toSet
-  )
-
-  override def listAccessPoliciesForUser(resource: Resource, user: WorkbenchUserId): Future[Set[AccessPolicy]] = Future {
-    val result = for {
-      entry <- Option(ldapConnectionPool.getEntry(subjectDn(user), Attr.memberOf))
-      members = getAttributes(entry, Attr.memberOf)
-      memberOfDns <- if(members.isEmpty) None else Some(members)
-    } yield {
-      val memberOfSubjects = memberOfDns.map(dnToSubject)
-      memberOfSubjects.collect {
-        case ripn@ResourceAndPolicyName(`resource`, _) => unmarshalAccessPolicy(ldapConnectionPool.getEntry(subjectDn(ripn)))
+  ))
+
+  override def listAccessPoliciesForUser(resource: ResourceIdentity, user: WorkbenchUserId): IO[Set[AccessPolicy]] = for {
+      entry <- cs.evalOn(blockingEc)(IO(ldapConnectionPool.getEntry(subjectDn(user), Attr.memberOf)))
+      members <- IO.pure(Option(entry).map(e => getAttributes(e, Attr.memberOf).toList))
+      accessPolicies <- members.traverse{
+        strs =>
+          val ripns = strs.mapFilter{
+            str =>
+              for{
+                subject <- Either.catchNonFatal(dnToSubject(str)).toOption
+                ripn <- subject match {
+                  case ripn: PolicyIdentity if ripn.resource.resourceId == resource.resourceId && ripn.resource.resourceTypeName == resource.resourceTypeName => Some(ripn)
+                  case _ => None
+                }
+              } yield ripn
+          }
+          val filters = ripns.grouped(batchSize).map(
+            batch => Filter.createORFilter(batch.map(r => Filter.createEqualityFilter(Attr.policy, r.accessPolicyName.value)).asJava)).toSeq
+          cs.evalOn(blockingEc)(IO(ldapSearchStream(resourceDn(resource), SearchScope.SUB, filters: _*)(unmarshalAccessPolicy).toSet))
       }
-    }
-    result.getOrElse(Set.empty)
-  }
+    } yield accessPolicies.getOrElse(Set.empty)
 
-  override def listFlattenedPolicyMembers(resourceAndPolicyName: ResourceAndPolicyName): Future[Set[WorkbenchUserId]] = Future {
-    ldapSearchStream(peopleOu, SearchScope.ONE, Filter.createEqualityFilter(Attr.memberOf, policyDn(resourceAndPolicyName)))(unmarshalUser).map(_.id).toSet
-  }
+  override def listFlattenedPolicyMembers(policyId: PolicyIdentity): IO[Set[WorkbenchUserId]] = cs.evalOn(blockingEc)(
+    IO(ldapSearchStream(peopleOu, SearchScope.ONE, Filter.createEqualityFilter(Attr.memberOf, policyDn(policyId)))(unmarshalUser).map(_.id).toSet)
```

Comment:
I believe this might be a good time to rework this query. @dvoet pointed out the same problem in the (soon to be deleted) function `listFlattenedGroupUsers` in `ldapDirectoryDAO` and the search I was doing to find intersection users in my PR. Check out the bottom row in the [LDAP knowledge share](https://broadinstitute.atlassian.net/wiki/spaces/GAWB/pages/622985222/LDAP). It could speed up the `listFlattenedPolicyMembers` query by a lot